### PR TITLE
Ignoring A ~ B constraints when A or B is non injective.

### PR DESCRIPTION
**Pull request by Gaspi**
Work on the #17  issue.
Ignoring all A ~ B constraints when A or B is non injective.

### CODE REVIEW

File: kernel/typing.ml [hunk 1]
```diff
@@ -135,11 +135,19 @@ let rec pseudo_u sg (sigma:SS.t) : (int*term*term) list -> SS.t option = functio
...
         | Kind, Kind | Type _, Type _ -> pseudo_u sg sigma lst
-        | DB (_,_,n), DB (_,_,n') when ( n=n' ) -> pseudo_u sg sigma lst
-        | Const (_,cst), Const (_,cst') when
-            ( name_eq cst cst' ) ->
-          pseudo_u sg sigma lst
-
+        | DB    (_,_,n), DB    (_,_,n') when n = n'           -> pseudo_u sg sigma lst
+        | Const (_,cst), Const (_,cst') when name_eq cst cst' -> pseudo_u sg sigma lst
+        | Const (l,cst), Const (l',cst') ->
+          if name_eq cst cst'
```
Comment (GuillaumeGen) [inline kernel/typing.ml]: Isn't this condition already checked one line upper ?

**Comment by GuillaumeGen** [comment]
I do not understand why if a Type symbol is definable we should accept the rule and otherwise we should refuse it.
Furthermore, it would introduce unspecified behavior, since (as far as I know) there isn't any article stating such a constraint (which again sounds very strange to me).

File: kernel/typing.ml (hunk 1 above)
Comment (Gaspi) [reply kernel/typing.ml]: My bad, I forgot to delete the line above. Fixed in the next commit.

**Comment by Gaspi** [comment]
This should probably be discussed during the next meeting...
The idea is that if no substitution allow for the left pattern to be well typed then it's (maybe) safe to accept the rule anyway. It's simply never going to be used.
Besides the rule might become well typed later when other rewrite rules are added to the system.

**Comment by GuillaumeGen** [comment]
I don't know if there is somewhere a definition of what is an accepted rewrite rule (probably called a well-typed rewrite rule, or the decidable under-approximation of it).
In the current state of Dedukti, your 3rd OK example is accepted, not the two first, which seems to be a weird behaviour (but maybe not).

**Comment by francoisthire** [comment]
We choose to merge #108 instead.

**Pull request closed without merging**